SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AutodiffClosureSpecialization.swift

```diff
@@ -0,0 +1,625 @@
+        return markReabstractionPartialApplyAsUsed(firstClosure: firstClosure, currentClosure: pai.callee, usedReabstractionClosures: &usedReabstractionClosures)
+    }
+
+    if let cvt = currentClosure as? ConvertFunctionInst {
+        usedReabstractionClosures.insert(cvt)
+        return markReabstractionPartialApplyAsUsed(firstClosure: firstClosure, currentClosure: cvt.fromFunction, usedReabstractionClosures: &usedReabstractionClosures)
+    }
+
+    if let cvt = currentClosure as? ConvertEscapeToNoEscapeInst {
+        usedReabstractionClosures.insert(cvt)
+        return markReabstractionPartialApplyAsUsed(firstClosure: firstClosure, currentClosure: cvt.fromFunction, usedReabstractionClosures: &usedReabstractionClosures)
+    }
+
+    if let mdi = currentClosure as? MarkDependenceInst {
+        usedReabstractionClosures.insert(mdi)
+        return markReabstractionPartialApplyAsUsed(firstClosure: firstClosure, currentClosure: mdi.value, usedReabstractionClosures: &usedReabstractionClosures)
+    }   
+
+    fatalError("Unexpected instruction used during closure reabstraction: \(currentClosure)")
+}
+
+// ===================== Types ===================== //
+
+/// Represents all the information required to
+/// represent a closure in isolation, i.e., outside of
+/// a callsite context where the closure may be getting
+/// passed as an argument.
+///
+/// Composed with other information inside a `ClosureArgDescriptor`
+/// to represent a closure as an argument at a callsite.
+class ClosureInfo {
+    let closure: SingleValueInstruction
+    var lifetimeFrontier: [Instruction]?
+
+    init(closure: SingleValueInstruction) {
+        self.closure = closure
+        self.lifetimeFrontier = nil
+    }
+
+    public func setLifetimeFrontier(_ frontier: [Instruction]) {
+        if self.lifetimeFrontier == nil {
+            self.lifetimeFrontier = frontier
+        } 
+    }
+}
+
+/// Represents a closure as an argument at a callsite.
+struct ClosureArgDescriptor {
+    let closureInfo: ClosureInfo
+    /// The index of the closure in the callsite's argument list.
+    let closureArgumentIndex: Int
+    let parameterInfo: ParameterInfo
+    /// This is only needed if we have guaranteed parameters. In most cases
+    /// it will have only one element, a return inst. 
+    var nonFailureExitBBs: [BasicBlock]
+}
+
+/// Represents a callsite containing one or more closure arguments.
+struct CallSiteDescriptor {
+    let applySite: ApplySite
+    var closureArgDescriptors: [ClosureArgDescriptor]
+    var silArgIndexToClosureArgDescIndex: [Int: Int]
+
+    public mutating func addClosureArgDescriptor(_ descriptor: ClosureArgDescriptor) {
+        self.silArgIndexToClosureArgDescIndex[descriptor.closureArgumentIndex] = self.closureArgDescriptors.count
+        self.closureArgDescriptors.append(descriptor)
+    }
+}
+
+/// A type capable of performing closure specialization on SIL functions
+protocol ClosureSpecializer {
```

Comment:
Why do we need this protocol?

Reply:
I'm anticipating that the closure-specializer for general Swift is going to diverge enough from the Autodiff specific closure-specializer. I can always remove the protocol and add it when I need it.

Reply:
Removed the protocol for now. Will add if needed when we incorporate general swift closure-specialization in here.